src/saveload/details.rs

```diff
@@ -0,0 +1,199 @@
+    type ReadStorages: SystemData<'a> + 'a;
+    /// Storages for write
+    type WriteStorages: SystemData<'a> + 'a;
+}
+
+/// This trait is implemented by any tuple where all elements are
+/// `Component + Serialize + DeserializeOwned`
+pub trait Components<M, E: Error>: for<'a> Storages<'a> {
+    /// Serializable and deserializable intermediate representation
+    type Data: Serialize + DeserializeOwned;
+
+    /// Saves `Component`s of entity into `Intermediate` serializable representation
+    fn save<'a, F>(
+        entity: Entity,
+        storages: &<Self as Storages<'a>>::ReadStorages,
+        ids: F,
+    ) -> Result<Self::Data, E>
+    where
+        F: FnMut(Entity) -> Option<M>;
+
+    /// Loads `Component`s to entity from `Intermediate` deserializable representation
+    fn load<'a, F>(
+        entity: Entity,
+        components: Self::Data,
+        storages: &mut <Self as Storages<'a>>::WriteStorages,
+        ids: F,
+    ) -> Result<(), E>
+    where
+        F: FnMut(M) -> Option<Entity>;
+}
+
+macro_rules! impl_components {
+    ($($a:ident|$b:ident),*) => {
+        impl<'a, $($a),*> Storages<'a> for ($($a,)*)
+            where $(
+                $a: Component,
+            )*
+        {
+            type ReadStorages = ($(ReadStorage<'a, $a>,)*);
+            type WriteStorages = ($(WriteStorage<'a, $a>,)*);
+        }
+
+        impl<M, E $(,$a)*> Components<M, E> for ($($a,)*)
+            where E: Error,
+            $(
+                $a: SaveLoadComponent<M>,
+                E: From<$a::Error>,
+            )*
+        {
+            type Data = ($(Option<$a::Data>,)*);
+
+            #[allow(unused_variables, unused_mut, non_snake_case)]
+            fn save<'a, F>(entity: Entity, storages: &($(ReadStorage<'a, $a>,)*), mut ids: F)
+                -> Result<($(Option<$a::Data>,)*), E>
+                where F: FnMut(Entity) -> Option<M>
+            {
+                let ($(ref $b,)*) = *storages;
+                Ok(($(
+                    $b.get(entity).map(|c| c.save(&mut ids).map(Some)).unwrap_or(Ok(None))?,
+                )*))
+            }
+
+            #[allow(unused_variables, unused_mut, non_snake_case)]
+            fn load<'a, F>(entity: Entity, components: ($(Option<$a::Data>,)*),
+                           storages: &mut ($(WriteStorage<'a, $a>,)*), mut ids: F)
+                -> Result<(), E>
+                where F: FnMut(M) -> Option<Entity>
+            {
+                let ($($a,)*) = components;
+                let ($(ref mut $b,)*) = *storages;
+                $(
+                    if let Some(a) = $a {
+                        $b.insert(entity, $a::load(a, &mut ids)?);
+                    } else {
+                        $b.remove(entity);
+                    }
+                )*
+                Ok(())
+            }
+        }
+
+        impl_components!(@ $($a|$b),*);
+    };
+
+    (@) => {};
+
+    (@ $ah:ident|$bh:ident $(,$a:ident|$b:ident)*) => {
```

Comment:
Could you explain that?

Reply:
This creates recursion. See it is called few lines above. It cuts head of list and call macro again.

Reply:
This make macro call itself with one less ident recursively

Reply:
Maybe add comment for explaining this.